README.md
[//]: # ( )
[//]: # (This file is automatically generated by a `metapak`)
[//]: # (module. Do not change it  except between the)
[//]: # (`content:start/end` flags, your changes would)
[//]: # (be overridden.)
[//]: # ( )
# swagger-http-router
> A HTTP router based on your Swagger/OpenAPI definition.

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/nfroidure/swagger-http-router/blob/master/LICENSE)
[![Build status](https://secure.travis-ci.org/nfroidure/swagger-http-router.svg)](https://travis-ci.org/nfroidure/swagger-http-router)
[![Coverage Status](https://coveralls.io/repos/nfroidure/swagger-http-router/badge.svg?branch=master)](https://coveralls.io/r/nfroidure/swagger-http-router?branch=master)
[![NPM version](https://badge.fury.io/js/swagger-http-router.svg)](https://npmjs.org/package/swagger-http-router)
[![Dependency Status](https://david-dm.org/nfroidure/swagger-http-router.svg)](https://david-dm.org/nfroidure/swagger-http-router)
[![devDependency Status](https://david-dm.org/nfroidure/swagger-http-router/dev-status.svg)](https://david-dm.org/nfroidure/swagger-http-router#info=devDependencies)
[![Package Quality](http://npm.packagequality.com/shield/swagger-http-router.svg)](http://packagequality.com/#?package=swagger-http-router)
[![Code Climate](https://codeclimate.com/github/nfroidure/swagger-http-router.svg)](https://codeclimate.com/github/nfroidure/swagger-http-router)


[//]: # (::contents:start)

Why write code when you have a Swagger/OpenAPI definition?

By taking part of the Swagger/OpenAPI standard and
 dependency injection  patterns, `swagger-http-router`
 provides a convenient, highly modular and easily
 testable REST tool.

## Usage
```js
import { constant } from 'knifecycle';
import initDB from './services/db';
import {
  initWepApplication
} from 'swagger-http-router';

import API from './swagger.api.json';
import * as HANDLERS from './handlers';

run();

async function run() {
  try {
    // STEP 1: Spawn a Knifecycle instance and attach
    // it the API definition and its handlers
    const $ = initWepApplication(API, HANDLERS);

    // STEP 2: Register additional services
    // Override the build in `uniqueId` service
    // with the UUID v4 function
    $.register(constant('uniqueId', uuid.v4))
    // Provide the process environment
    .register(constant('ENV', process.env))
    // Register the database initializer
    .register(initDB);

    // STEP 3: Run your app!
    // Run the execution silo that encapsulates the app
    // Note that the `httpServer` and `process` services
    // are injected for their respective side effects:
    // creating the server and managing the process
    // lifecycle
    const { ENV, log, $destroy } = await $.run(['ENV', 'log', 'httpServer', 'process', '$destroy']);

    log('info', `On air 🚀🌕`);

    if(ENV.DRY_RUN) {
      await $destroy();
    }
  } catch(err) {
    console.error('💀 - Cannot launch the process:', err.stack);
    process.exit(1);
  }
)
```

In order to work, your Swagger definition endpoints
must provide an
 [`operationId`](http://swagger.io/specification/#operationObject).
 This is how the router figures out which handler
 to run. Those ids have to be unique. Here is
 a sample Swagger definition you could use as is:
```js
// file: ./my-swagger.json
{
  "host": "localhost:1337",
  "basePath": "/v1",
  "schemes": ["http"],
  // (...)
  "paths": {
    "GET": {
      "/users/{userId}": {
        "operationId": "getUser",
        "summary": "Retrieve a user.",
        "produces": [
          "application/json"
        ],
        "parameters": [{
          "in": "path",
          "name": "userId",
          "type": "number",
          "pattern": "^[0-9]+$"
        }, {
          "in": "query",
          "name": "extended",
          "type": "boolean"
        }, {
          "in": "header",
          "name": "Content-Type",
          "type": "string"
        }],
        "responses": {
          "200": {
            "description": "User found",
            "schema": {
              "type": "object",
              "properties": {
                "id": { "type": "number" },
                "name": { "type": "string" }
              }
            }
          },
          "404": {
            "description": "User not found"
          }
        }
      }
    }
  }
}
```

To bring to the router the logic that each
 endpoint implements, you have to create
 handlers for each `operationId`:
```js
// file: ./handlers.js

// Knifecycle is the dependency injection tool
// we use. It provides decorators to declare
// which dependencies to inject in your handlers
import { initializer } from 'knifecycle/dist/util';

export default initializer(
  {
    name: 'getUser',
    type: 'service',
    inject: ['db'],
  },
  getUser
);

async function getUser({ db }) {
  return async ({ userId }) => {
    const user = await db.query('users', {
      id: userId,
    });

    return {
      status: 200,
      headers: {},
      body: {
        id: userId,
        name: user.name,
      }
    };
  }
}
```

As you can see, handlers are just asynchronous functions
that takes the request parameters in input and provide
a JSON serializable response in output.

This router is designed to be used with a DI system and
 is particularly useful with
 [`knifecycle`](https://github.com/nfroidure/knifecycle).

That said, you could completely avoid using a DI system
 by simply using the initializers as functions and handle
 their initialization manually. See this
  [alternate example](https://gist.github.com/nfroidure/647189bdeffef33bced3a3b6d924640e).

## Goal

This router is just my way to do things. It is nice
 if you use it and I'd be happy if you improve it.

To be honest, I think this is a better approach but I do
 not want to spend energy and fight with giants to make
 this a standard approach. It means that it will probably
 never be the next hype and if you use it you must feel
 confident with forking and maintaining it yourself.
 That said, the code is well documented and not that hard.
 Also, the handlers you will end with will be easily
 reusable in any framework with little to no changes.

You may wonder why I found that I'd better write
 my own router instead of using current solutions
 like `ExpressJS` or `HapiJS`:
- I want documentation first APIs. No documentation, no
 web service.
- I want my code to be clear and descriptive instead of
 binded to some cryptic middleware or plugin defined
 elsewhere. Here are some
 [thoughts on middlewares](http://insertafter.com/en/blog/no_more_middlewares.html)
 that explain this statement in more depth.
 - I want easily testable and reusable handlers just
  returning plain JSON. To be able to reuse it in
  multiple situations: a lambda/serverless back-end,
  when rendering server side React views or in my
  GraphQL server resolvers.
- I prefer functional programming: it just makes my code
 better. There are too many encapsulated states in existing
 frameworks. I just want my handlers to be pure and
 composable. For example, why adding a CORS middleware or
 plugin when you can just compose handlers?
```js
import { reuseSpecialProps } from 'knifecycle/dist/util';

const CORS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Keep-Alive,User-Agent',
};

export function wrapWithCORS(initHandler) {
  // `reuseSpecialProps` create a new initializer
  // with the original initializer properties
  // applyed on it.
  return reuseSpecialProps(
    initHandler,
    initHandlerWithCORS.bind(null, initHandler)
  );
}

// This function is the actual initializer that
// wraps the handler initializer. It is executed
// once at startup.
async function initHandlerWithCORS(initHandler, services) => {
  const handler = await initHandler(services);

  return handleWithCors.bind(null, handler);
}

// And finally this one applies CORS to the
// response
async function handleWithCors(handler, parameters) {
  const response = await handler(parameters);

  return  {
    ...response,
    headers: {
      ...response.headers,
      ...CORS,
    }
  };
}

```
- and finally, I want to be able to instrument my code
 without having to do ugly hacks. This is why DI and
 Inversion of Control are at the core of my way to
 handle web services.

You may want to have a look at the
 [architecture notes](./ARCHITECTURE.md) of this module
 to better grasp how it is built.

## Recommendations

The above usage section shows you a very basic
 usage of this router. For larger apps:
- you may want to build you Swagger file to avoid
 repeating yourself. It won't change anything for
 `swagger-http-router` since it just assumes a
 Swagger file.
- you will probably end up by automating the
 handlers loading with a configuration file.
 At that point, the DI system will become very
 handy.
- you will certainly need some more services
 to make your app work. Please double check if
 one exists before creating your own. Also,
 handlers can be reused so feel free to
 publish yours and add your Swagger path
 objects to them in order for your users to
 add them to their own Swagger build.

# API
## Functions

<dl>
<dt><a href="#initErrorHandler">initErrorHandler(services)</a> ⇒ <code>Promise</code></dt>
<dd><p>Initialize an error handler for the
HTTP router</p>
</dd>
<dt><a href="#initWepApplication">initWepApplication(API, HANDLERS, [$])</a> ⇒ <code>Knifecycle</code></dt>
<dd><p>Initialize a web application</p>
</dd>
<dt><a href="#registerHandlers">registerHandlers($, HANDLERS)</a> ⇒ <code>void</code></dt>
<dd><p>Register the handlers hash into the given Knifecycle
 instance</p>
</dd>
<dt><a href="#initHTTPRouter">initHTTPRouter(services)</a> ⇒ <code>Promise</code></dt>
<dd><p>Initialize an HTTP router</p>
</dd>
<dt><a href="#initHTTPServer">initHTTPServer(services)</a> ⇒ <code>Promise</code></dt>
<dd><p>Initialize an HTTP server</p>
</dd>
<dt><a href="#initHTTPTransaction">initHTTPTransaction(services)</a> ⇒ <code>Promise.&lt;function()&gt;</code></dt>
<dd><p>Instantiate the httpTransaction service</p>
</dd>
<dt><a href="#flattenSwagger">flattenSwagger(API)</a> ⇒ <code>Object</code></dt>
<dd><p>Flatten the inputed Swagger file
 object</p>
</dd>
<dt><a href="#getSwaggerOperations">getSwaggerOperations(API)</a> ⇒ <code>Array</code></dt>
<dd><p>Return a Swagger operation in a more
 convenient way to iterate onto its
 operations</p>
</dd>
</dl>

<a name="initErrorHandler"></a>

## initErrorHandler(services) ⇒ <code>Promise</code>
Initialize an error handler for the
HTTP router

**Kind**: global function  
**Returns**: <code>Promise</code> - A promise of a function to handle errors  

| Param | Type | Description |
| --- | --- | --- |
| services | <code>Object</code> | The services the server depends on |
| [services.ENV] | <code>Object</code> | The services the server depends on |
| [services.DEBUG_NODE_ENVS] | <code>Array</code> | The environnement that activate debugging  (prints stack trace in HTTP errors responses) |
| [services.STRINGIFYERS] | <code>Object</code> | The synchronous body stringifyers |

<a name="initErrorHandler..errorHandler"></a>

### initErrorHandler~errorHandler(transactionId, responseSpec, err) ⇒ <code>Promise</code>
Handle an HTTP transaction error an
map it to a serializable response

**Kind**: inner method of [<code>initErrorHandler</code>](#initErrorHandler)  
**Returns**: <code>Promise</code> - A promise resolving when the operation
 completes  

| Param | Type | Description |
| --- | --- | --- |
| transactionId | <code>String</code> | A raw NodeJS HTTP incoming message |
| responseSpec | <code>Object</code> | The response specification |
| err | <code>HTTPError</code> | The encountered error |

<a name="initWepApplication"></a>

## initWepApplication(API, HANDLERS, [$]) ⇒ <code>Knifecycle</code>
Initialize a web application

**Kind**: global function  
**Returns**: <code>Knifecycle</code> - The passed in Knifecycle instance or the one created
 by default.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| API | <code>Object</code> |  | The Swagger definition of the we application |
| HANDLERS | <code>Object</code> |  | The handlers for each operations defined by the  Swagger definition. |
| [$] | <code>Knifecycle</code> | <code>getInstance(</code> | A Knifecycle instance on which to set the application  up. |

<a name="registerHandlers"></a>

## registerHandlers($, HANDLERS) ⇒ <code>void</code>
Register the handlers hash into the given Knifecycle
 instance

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| $ | <code>Knifecycle</code> | The Knifecycle instance on which to set up the handlers |
| HANDLERS | <code>Object</code> | The handlers hash |

<a name="initHTTPRouter"></a>

## initHTTPRouter(services) ⇒ <code>Promise</code>
Initialize an HTTP router

**Kind**: global function  
**Returns**: <code>Promise</code> - A promise of a function to handle HTTP requests.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services the server depends on |
| services.API | <code>Object</code> |  | The Swagger definition of the API |
| services.HANDLERS | <code>Object</code> |  | The handlers for the operations decribe  by the Swagger API definition |
| [services.ENV] | <code>Object</code> |  | The services the server depends on |
| [services.DEBUG_NODE_ENVS] | <code>Array</code> |  | The environnement that activate debugging  (prints stack trace in HTTP errors responses) |
| [services.BUFFER_LIMIT] | <code>String</code> |  | The maximum bufferisation before parsing the  request body |
| [services.PARSERS] | <code>Object</code> |  | The synchronous body parsers (for operations  that defines a request body schema) |
| [services.STRINGIFYERS] | <code>Object</code> |  | The synchronous body stringifyers (for  operations that defines a response body  schema) |
| [services.log] | <code>function</code> | <code>noop</code> | A logging function |
| services.httpTransaction | <code>function</code> |  | A function to create a new HTTP transaction |

<a name="initHTTPServer"></a>

## initHTTPServer(services) ⇒ <code>Promise</code>
Initialize an HTTP server

**Kind**: global function  
**Returns**: <code>Promise</code> - A promise of an object with a NodeJS HTTP server
 in its `service` property.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services the server depends on |
| services.ENV | <code>Object</code> |  | The process environment variables |
| services.httpRouter | <code>function</code> |  | The function to run with the req/res tuple |
| [services.log] | <code>function</code> | <code>noop</code> | A logging function |

<a name="initHTTPTransaction"></a>

## initHTTPTransaction(services) ⇒ <code>Promise.&lt;function()&gt;</code>
Instantiate the httpTransaction service

**Kind**: global function  
**Returns**: <code>Promise.&lt;function()&gt;</code> - A promise of the httpTransaction function  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services to inject |
| [services.TIMEOUT] | <code>Number</code> | <code>30000</code> | A number indicating how many ms the transaction  should take to complete before being cancelled. |
| [services.TRANSACTIONS] | <code>Object</code> | <code>{}</code> | A hash of every current transactions |
| services.time | <code>function</code> |  | A timing function |
| services.delay | <code>Object</code> |  | A delaying service |
| [services.log] | <code>function</code> |  | A logging function |
| [services.uniqueId] | <code>function</code> |  | A function returning unique identifiers |

**Example**  
```js
import { initHTTPTransaction } from 'swagger-http-router';

const httpTransaction = await initHTTPTransaction({
  log: console.log.bind(console),
  time: Date.now.bind(Date),
});
```
<a name="initHTTPTransaction..httpTransaction"></a>

### initHTTPTransaction~httpTransaction(req, res) ⇒ <code>Array</code>
Create a new HTTP transaction

**Kind**: inner method of [<code>initHTTPTransaction</code>](#initHTTPTransaction)  
**Returns**: <code>Array</code> - The normalized request and the HTTP
transaction created in an array.  

| Param | Type | Description |
| --- | --- | --- |
| req | <code>HTTPRequest</code> | A raw NodeJS HTTP incoming message |
| res | <code>HTTPResponse</code> | A raw NodeJS HTTP response |

<a name="flattenSwagger"></a>

## flattenSwagger(API) ⇒ <code>Object</code>
Flatten the inputed Swagger file
 object

**Kind**: global function  
**Returns**: <code>Object</code> - The flattened Swagger definition  

| Param | Type | Description |
| --- | --- | --- |
| API | <code>Object</code> | An Object containing a parser Swagger JSON |

<a name="getSwaggerOperations"></a>

## getSwaggerOperations(API) ⇒ <code>Array</code>
Return a Swagger operation in a more
 convenient way to iterate onto its
 operations

**Kind**: global function  
**Returns**: <code>Array</code> - An array of all the Swagger operations  

| Param | Type | Description |
| --- | --- | --- |
| API | <code>Object</code> | The flattened Swagger defition |

**Example**  
```js
getSwaggerOperations(API)
.map((operation) => {
   const { path, method, operationId, parameters } = operation;

  // Do something with that operation
});
```

[//]: # (::contents:end)

# API
## Functions

<dl>
<dt><a href="#initErrorHandler">initErrorHandler(services)</a> ⇒ <code>Promise</code></dt>
<dd><p>Initialize an error handler for the
HTTP router</p>
</dd>
<dt><a href="#initWepApplication">initWepApplication(API, HANDLERS, [$])</a> ⇒ <code>Knifecycle</code></dt>
<dd><p>Initialize a web application</p>
</dd>
<dt><a href="#initHTTPRouter">initHTTPRouter(services)</a> ⇒ <code>Promise</code></dt>
<dd><p>Initialize an HTTP router</p>
</dd>
<dt><a href="#initHTTPServer">initHTTPServer(services)</a> ⇒ <code>Promise</code></dt>
<dd><p>Initialize an HTTP server</p>
</dd>
<dt><a href="#initHTTPTransaction">initHTTPTransaction(services)</a> ⇒ <code>Promise.&lt;function()&gt;</code></dt>
<dd><p>Instantiate the httpTransaction service</p>
</dd>
<dt><a href="#flattenSwagger">flattenSwagger(API)</a> ⇒ <code>Object</code></dt>
<dd><p>Flatten the inputed Swagger file
 object</p>
</dd>
<dt><a href="#getSwaggerOperations">getSwaggerOperations(API)</a> ⇒ <code>Array</code></dt>
<dd><p>Return a Swagger operation in a more
 convenient way to iterate onto its
 operations</p>
</dd>
</dl>

<a name="initErrorHandler"></a>

## initErrorHandler(services) ⇒ <code>Promise</code>
Initialize an error handler for the
HTTP router

**Kind**: global function  
**Returns**: <code>Promise</code> - A promise of a function to handle errors  

| Param | Type | Description |
| --- | --- | --- |
| services | <code>Object</code> | The services the server depends on |
| [services.ENV] | <code>Object</code> | The services the server depends on |
| [services.DEBUG_NODE_ENVS] | <code>Array</code> | The environnement that activate debugging  (prints stack trace in HTTP errors responses) |
| [services.STRINGIFYERS] | <code>Object</code> | The synchronous body stringifyers |

<a name="initErrorHandler..errorHandler"></a>

### initErrorHandler~errorHandler(transactionId, responseSpec, err) ⇒ <code>Promise</code>
Handle an HTTP transaction error and
map it to a serializable response

**Kind**: inner method of [<code>initErrorHandler</code>](#initErrorHandler)  
**Returns**: <code>Promise</code> - A promise resolving when the operation
 completes  

| Param | Type | Description |
| --- | --- | --- |
| transactionId | <code>String</code> | A raw NodeJS HTTP incoming message |
| responseSpec | <code>Object</code> | The response specification |
| err | <code>HTTPError</code> | The encountered error |

<a name="initWepApplication"></a>

## initWepApplication(API, HANDLERS, [$]) ⇒ <code>Knifecycle</code>
Initialize a web application

**Kind**: global function  
**Returns**: <code>Knifecycle</code> - The passed in Knifecycle instance or the one created
 by default.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| API | <code>Object</code> |  | The Swagger definition of the we application |
| HANDLERS | <code>Object</code> |  | The handlers for each operations defined by the  Swagger definition. |
| [$] | <code>Knifecycle</code> | <code>getInstance(</code> | A Knifecycle instance on which to set the application  up. |

<a name="initHTTPRouter"></a>

## initHTTPRouter(services) ⇒ <code>Promise</code>
Initialize an HTTP router

**Kind**: global function  
**Returns**: <code>Promise</code> - A promise of a function to handle HTTP requests.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services the server depends on |
| services.API | <code>Object</code> |  | The Swagger definition of the API |
| services.HANDLERS | <code>Object</code> |  | The handlers for the operations decribe  by the Swagger API definition |
| [services.ENV] | <code>Object</code> |  | The services the server depends on |
| [services.DEBUG_NODE_ENVS] | <code>Array</code> |  | The environnement that activate debugging  (prints stack trace in HTTP errors responses) |
| [services.BUFFER_LIMIT] | <code>String</code> |  | The maximum bufferisation before parsing the  request body |
| [services.PARSERS] | <code>Object</code> |  | The synchronous body parsers (for operations  that defines a request body schema) |
| [services.STRINGIFYERS] | <code>Object</code> |  | The synchronous body stringifyers (for  operations that defines a response body  schema) |
| [services.ENCODERS] | <code>Object</code> |  | A map of encoder stream constructors |
| [services.DECODERS] | <code>Object</code> |  | A map of decoder stream constructors |
| [services.QUERY_PARSER] | <code>Object</code> |  | A query parser with the `strict-qs` signature |
| [services.log] | <code>function</code> | <code>noop</code> | A logging function |
| services.httpTransaction | <code>function</code> |  | A function to create a new HTTP transaction |

<a name="initHTTPServer"></a>

## initHTTPServer(services) ⇒ <code>Promise</code>
Initialize an HTTP server

**Kind**: global function  
**Returns**: <code>Promise</code> - A promise of an object with a NodeJS HTTP server
 in its `service` property.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services the server depends on |
| services.ENV | <code>Object</code> |  | The process environment variables |
| services.HOST | <code>Object</code> |  | The server host |
| services.PORT | <code>Object</code> |  | The server port |
| services.MAX_HEADERS_COUNT | <code>Object</code> |  | The https://nodejs.org/api/http.html#http_server_maxheaderscount |
| services.KEEP_ALIVE_TIMEOUT | <code>Object</code> |  | See https://nodejs.org/api/http.html#http_server_keepalivetimeout |
| services.MAX_CONNECTIONS | <code>Object</code> |  | See https://nodejs.org/api/net.html#net_server_maxconnections |
| services.TIMEOUT | <code>Object</code> |  | See https://nodejs.org/api/http.html#http_server_timeout |
| services.httpRouter | <code>function</code> |  | The function to run with the req/res tuple |
| [services.log] | <code>function</code> | <code>noop</code> | A logging function |

<a name="initHTTPTransaction"></a>

## initHTTPTransaction(services) ⇒ <code>Promise.&lt;function()&gt;</code>
Instantiate the httpTransaction service

**Kind**: global function  
**Returns**: <code>Promise.&lt;function()&gt;</code> - A promise of the httpTransaction function  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services to inject |
| [services.TIMEOUT] | <code>Number</code> | <code>30000</code> | A number indicating how many ms the transaction  should take to complete before being cancelled. |
| [services.TRANSACTIONS] | <code>Object</code> | <code>{}</code> | A hash of every current transactions |
| services.time | <code>function</code> |  | A timing function |
| services.delay | <code>Object</code> |  | A delaying service |
| [services.log] | <code>function</code> |  | A logging function |
| [services.uniqueId] | <code>function</code> |  | A function returning unique identifiers |

**Example**  
```js
import { initHTTPTransaction } from 'swagger-http-router';

const httpTransaction = await initHTTPTransaction({
  log: console.log.bind(console),
  time: Date.now.bind(Date),
});
```
<a name="initHTTPTransaction..httpTransaction"></a>

### initHTTPTransaction~httpTransaction(req, res) ⇒ <code>Array</code>
Create a new HTTP transaction

**Kind**: inner method of [<code>initHTTPTransaction</code>](#initHTTPTransaction)  
**Returns**: <code>Array</code> - The normalized request and the HTTP
transaction created in an array.  

| Param | Type | Description |
| --- | --- | --- |
| req | <code>HTTPRequest</code> | A raw NodeJS HTTP incoming message |
| res | <code>HTTPResponse</code> | A raw NodeJS HTTP response |

<a name="flattenSwagger"></a>


CHANGELOG.md
## [4.2.1](https://github.com/nfroidure/swagger-http-router/compare/v4.2.0...v4.2.1) (2019-02-02)



# [4.2.0](https://github.com/nfroidure/swagger-http-router/compare/v4.1.0...v4.2.0) (2018-11-22)


### Bug Fixes

* **Tests:** Ensure that body checks reflect real world usage ([70f2fe6](https://github.com/nfroidure/swagger-http-router/commit/70f2fe6))



# [4.1.0](https://github.com/nfroidure/swagger-http-router/compare/v4.0.0...v4.1.0) (2018-11-11)


### Bug Fixes

* **httpRouter:** Avoid mutating the injected API object ([7ff3eb8](https://github.com/nfroidure/swagger-http-router/commit/7ff3eb8)), closes [#9](https://github.com/nfroidure/swagger-http-router/issues/9)


### Features

* **httpRouter:** Add some more check to the API definition ([d295b95](https://github.com/nfroidure/swagger-http-router/commit/d295b95)), closes [#6](https://github.com/nfroidure/swagger-http-router/issues/6) [#7](https://github.com/nfroidure/swagger-http-router/issues/7)



# [4.0.0](https://github.com/nfroidure/swagger-http-router/compare/v3.0.0...v4.0.0) (2018-11-04)


### chore

* **Dependencies:** Update dependencies ([e34c7f8](https://github.com/nfroidure/swagger-http-router/commit/e34c7f8))


### BREAKING CHANGES

* **Dependencies:** Drop support for Node < 8



<a name="3.0.0"></a>
# [3.0.0](https://github.com/nfroidure/swagger-http-router/compare/v2.0.0...v3.0.0) (2018-10-21)


### Code Refactoring

* **HTTPServer:** Avoid picking HOST/PORT in ENV ([58beb67](https://github.com/nfroidure/swagger-http-router/commit/58beb67)), closes [#8](https://github.com/nfroidure/swagger-http-router/issues/8)


### Features

* **httpServer:** Add new HTTP server options ([cfb01c5](https://github.com/nfroidure/swagger-http-router/commit/cfb01c5))


### BREAKING CHANGES

* **HTTPServer:** The HOST/PORT constants must now be set instead of setting it to ENV.



<a name="2.0.0"></a>
# [2.0.0](https://github.com/nfroidure/swagger-http-router/compare/v1.6.3...v2.0.0) (2018-10-14)


### Bug Fixes

* **httpTransaction:** Repair error casting when YHTTPError instances are different ([ed99b8f](https://github.com/nfroidure/swagger-http-router/commit/ed99b8f))



<a name="1.6.3"></a>
## [1.6.3](https://github.com/nfroidure/swagger-http-router/compare/v1.6.2...v1.6.3) (2018-09-16)


### Bug Fixes

* **Tests:** Fix meaningless failures between Node versions ([abab401](https://github.com/nfroidure/swagger-http-router/commit/abab401))



<a name="1.6.2"></a>
## [1.6.2](https://github.com/nfroidure/swagger-http-router/compare/v1.6.1...v1.6.2) (2018-05-17)


### Bug Fixes

* **Build:** Fix the Babel build for AWS Lambdas ([8be86c1](https://github.com/nfroidure/swagger-http-router/commit/8be86c1))



<a name="1.6.1"></a>
## [1.6.1](https://github.com/nfroidure/swagger-http-router/compare/v1.6.0...v1.6.1) (2018-05-06)



<a name="1.6.0"></a>
# [1.6.0](https://github.com/nfroidure/swagger-http-router/compare/v1.5.3...v1.6.0) (2018-04-11)


### Features

* **HTTPRouter:** Allow to inject a different query parser ([d00b712](https://github.com/nfroidure/swagger-http-router/commit/d00b712)), closes [#5](https://github.com/nfroidure/swagger-http-router/issues/5)
* **HTTPServer:** Allow to destroy sockets for a quicker shutdown ([9467680](https://github.com/nfroidure/swagger-http-router/commit/9467680)), closes [#4](https://github.com/nfroidure/swagger-http-router/issues/4)


### Performance Improvements

* **HTTPServer:** Avoid keeping track of opened sockets when unuseful ([2d15098](https://github.com/nfroidure/swagger-http-router/commit/2d15098))



<a name="1.5.3"></a>
## [1.5.3](https://github.com/nfroidure/swagger-http-router/compare/v1.5.2...v1.5.3) (2018-03-18)


### Bug Fixes

* **NPM:** Fix main file ([e33a5a8](https://github.com/nfroidure/swagger-http-router/commit/e33a5a8))



<a name="1.5.2"></a>
## [1.5.2](https://github.com/nfroidure/swagger-http-router/compare/v1.5.1...v1.5.2) (2018-03-18)



<a name="1.5.1"></a>
## [1.5.1](https://github.com/nfroidure/swagger-http-router/compare/v1.5.0...v1.5.1) (2018-03-18)


### Bug Fixes

* **HTTPTransaction:** Prevent conflicts for transaction ids ([a6ac5c5](https://github.com/nfroidure/swagger-http-router/commit/a6ac5c5))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/nfroidure/swagger-http-router/compare/v1.4.0...v1.5.0) (2018-03-18)



<a name="1.4.0"></a>
# [1.4.0](https://github.com/nfroidure/swagger-http-router/compare/v1.3.0...v1.4.0) (2018-03-07)


### Features

* **PARSERS:** Allow async parsers (for multipart/form-data support) ([3ddd90c](https://github.com/nfroidure/swagger-http-router/commit/3ddd90c))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/nfroidure/swagger-http-router/compare/v1.2.1...v1.3.0) (2017-11-12)


### Features

* **errorHandler:** Inject and expose the errorHandler ([5b905a7](https://github.com/nfroidure/swagger-http-router/commit/5b905a7)), closes [#2](https://github.com/nfroidure/swagger-http-router/issues/2)
* **PARSERS/STRINGIFYERS:** Add `application/x-www-form-urlencoded` support ([8afd402](https://github.com/nfroidure/swagger-http-router/commit/8afd402))



<a name="1.2.1"></a>
## [1.2.1](https://github.com/nfroidure/swagger-http-router/compare/v1.2.0...v1.2.1) (2017-09-11)


### Bug Fixes

* **Router:** Take in count user agents that sends charsets capitalized ([86d719f](https://github.com/nfroidure/swagger-http-router/commit/86d719f))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/nfroidure/swagger-http-router/compare/v1.1.0...v1.2.0) (2017-09-11)


### Features

* **Router:** Add a way to set headers for errors ([a84ff5a](https://github.com/nfroidure/swagger-http-router/commit/a84ff5a))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/nfroidure/swagger-http-router/compare/v1.0.1...v1.1.0) (2017-08-06)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/nfroidure/swagger-http-router/compare/v1.0.0...v1.0.1) (2017-08-02)


### Bug Fixes


src/index.mocha.js
/* eslint max-nested-callbacks: 0 */
'use strict';

const sinon = require('sinon');
const assert = require('assert');
const supertest = require('supertest');

const { initializer, Knifecycle } = require('knifecycle');
const { initWepApplication } = require('../src');

const API = require('../fixtures/swagger.api.json');
const HANDLERS = {
  ping: initializer(
    {
      name: 'ping',
      type: 'service',
    },
    () =>
      Promise.resolve(() => ({
        status: 200,
      }))
  ),
  headUserAvatar: initializer(
    {
      name: 'headUserAvatar',
      type: 'service',
    },
    () =>
      Promise.resolve(() => ({
        status: 200,
      }))
  ),
  getUserAvatar: initializer(
    {
      name: 'getUserAvatar',
      type: 'service',
    },
    () =>
      Promise.resolve(() => ({
        status: 200,
      }))
  ),
  putUserAvatar: initializer(
    {
      name: 'putUserAvatar',
      type: 'service',
    },
    () =>
      Promise.resolve(() => ({
        status: 200,
      }))
  ),
  deleteUserAvatar: initializer(
    {
      name: 'deleteUserAvatar',
      type: 'service',
    },
    () =>
      Promise.resolve(() => ({
        status: 200,
      }))
  ),
  getUser: initializer(
    {
      name: 'getUser',
      type: 'service',
      inject: ['db'],
    },
    ({ db }) =>
      Promise.resolve(({ userId }) =>
        db
          .query('users', {
            id: userId,
          })
          .then(user => ({
            status: 200,
            headers: {},
            body: {
              id: userId,
              name: user.name,
            },
          }))
      )
  ),
  putUser: initializer(
    {
      name: 'putUser',
      type: 'service',
      inject: ['db'],
    },
    ({ db }) =>
      Promise.resolve(({ userId }) =>
        db
          .query('users', {
            id: userId,
          })
          .then(user => ({
            status: 200,
            headers: {},
            body: {
              id: userId,
              name: user.name,
            },
          }))
      )
  ),
};

describe('initWepApplication', () => {
  let $;

  beforeEach(() => {
    $ = new Knifecycle();
    initWepApplication(API, HANDLERS, $);
    $.constant('db', { query: sinon.stub() });
    $.constant('log', sinon.stub());
    $.constant('ENV', { PORT: 1664, HOST: 'localhost' });
    $.constant('time', sinon.stub().returns(1337));
  });

  describe('with a few routes', () => {
    it('should work as expected', done => {
      $.run(['ENV', 'log', 'db', 'httpServer', '$destroy'])
        .then(({ ENV, log, db, $destroy }) => {
          db.query.returns(
            Promise.resolve({
              name: 'John Doe',
            })
          );
          return new Promise(resolve => {
            supertest(`http://${ENV.HOST}:${ENV.PORT}`)
              .get('/v1/users/1?extended=false')
              .unset('User-Agent')
              .expect(200)
              .end((err, res) => {
                if (err) {
                  console.log('Logs:', log.args); // eslint-disable-line
                  done(err);
                  return;
                }
                assert.deepEqual(db.query.args, [['users', { id: 1 }]]);
                assert.deepEqual(res.body, {
                  id: 1,
                  name: 'John Doe',
                });
                resolve($destroy());
              });
          }).then(() => {
            assert.deepEqual(log.args, [
              ['debug', 'Delay service initialized.'],
              ['debug', 'HTTP Transaction initialized.'],
              ['debug', 'HTTP Router initialized.'],
              ['info', 'HTTP Server listening at "http://localhost:1664".'],
              ['debug', 'Created a delay:', 30000],
              [
                'info',
                {
                  endInBytes: 116,
                  endOutBytes: 191,
                  endTime: 1337,
                  id: '0',
                  ip: '127.0.0.1',
                  method: 'GET',
                  protocol: 'http',
                  reqHeaders: {
                    'accept-encoding': 'gzip, deflate',
                    connection: 'close',
                    host: 'localhost:1664',
                  },
                  resHeaders: {
                    'content-type': 'application/json',
                  },
                  startInBytes: 116,
                  startOutBytes: 0,
                  startTime: 1337,
                  statusCode: 200,
                  url: '/v1/users/1?extended=false',
                  errored: false,
                },
              ],
              ['debug', 'Cleared a delay'],
              ['debug', 'Closing HTTP server.'],
              ['debug', 'HTTP server closed'],
            ]);
          });
        })
        .then(() => done())
        .catch(done);
    });
  });
});
